File: src/core/semver.ts
/**
 * node-semver 부분집합 — 기획서 4.3 / 11.3
 *
 * 지원: 비교자 `>= > <= < =` + 공백 AND 결합 (`>=3.2.0 <3.3.0`).
 * 미지원(파싱 거부): `||`(OR) · `^` · `~` · x-range · hyphen-range.
 * → 범위를 항상 명시적 하한·상한으로 강제해 충돌 검사(8.2)를 단순·안전하게.
 *
 * 프리릴리스(`3.2.0-rc1`)는 기본 미매칭, matchPrerelease=true일 때만 평가(11.3).
 */

export interface SemVer {
  readonly major: number;
  readonly minor: number;
  readonly patch: number;
  /** 프리릴리스 식별자 배열(비어있으면 정식 릴리스). 예: 3.2.0-rc.1 → ["rc", 1] */
  readonly prerelease: readonly (string | number)[];
}

const CORE = /^(\d+)\.(\d+)\.(\d+)(?:-([0-9A-Za-z.-]+))?(?:\+[0-9A-Za-z.-]+)?$/;

export function parseVersion(input: string): SemVer {
  const m = CORE.exec(input.trim());
  if (!m) throw new Error(`유효하지 않은 버전: "${input}"`);
  const pre = m[4];
  const prerelease =
    pre === undefined
      ? []
      : pre.split(".").map((id) => (/^\d+$/.test(id) ? Number(id) : id));
  return { major: Number(m[1]), minor: Number(m[2]), patch: Number(m[3]), prerelease };
}

export function isPrerelease(v: SemVer): boolean {
  return v.prerelease.length > 0;
}

/** semver 우선순위 비교: a<b → -1, a>b → 1, 동일 → 0 (프리릴리스 규칙 포함). */
export function compare(a: SemVer, b: SemVer): number {
  if (a.major !== b.major) return a.major < b.major ? -1 : 1;
  if (a.minor !== b.minor) return a.minor < b.minor ? -1 : 1;
  if (a.patch !== b.patch) return a.patch < b.patch ? -1 : 1;
  // 프리릴리스가 있는 쪽이 낮다.
  const ap = a.prerelease, bp = b.prerelease;
  if (ap.length === 0 && bp.length === 0) return 0;
  if (ap.length === 0) return 1;
  if (bp.length === 0) return -1;
  const n = Math.min(ap.length, bp.length);
  for (let i = 0; i < n; i++) {
    const x = ap[i]!, y = bp[i]!;
    const xn = typeof x === "number", yn = typeof y === "number";
    if (xn && yn) { if (x !== y) return (x as number) < (y as number) ? -1 : 1; }
    else if (xn) return -1; // 숫자 식별자는 비숫자보다 낮다
    else if (yn) return 1;
    else if (x !== y) return (x as string) < (y as string) ? -1 : 1;
  }
  if (ap.length !== bp.length) return ap.length < bp.length ? -1 : 1;
  return 0;
}

export type Op = ">=" | "<=" | ">" | "<" | "=";
export interface Comparator {
  readonly op: Op;
  readonly version: SemVer;
}

const REJECT = [
  { pat: "||", why: "OR(||)는 미지원 — 명시적 하한·상한만 허용" },
  { pat: "^", why: "캐럿(^) 범위는 미지원 — 명시적 비교자로 표기" },
  { pat: "~", why: "틸드(~) 범위는 미지원 — 명시적 비교자로 표기" },
  { pat: " - ", why: "hyphen-range는 미지원 — 명시적 비교자로 표기" },
];

const COMPARATOR = /^(>=|<=|>|<|=)?(\d+\.\d+\.\d+(?:-[0-9A-Za-z.-]+)?)$/;

/**
 * 부분집합 범위식을 Comparator[]로 파싱. 미지원 문법은 즉시 예외(11.3 '파싱 거부').
 * 반환된 비교자들은 AND로 결합된다.
 */
export function parseRange(input: string): Comparator[] {
  const raw = input.trim();
  if (raw === "") throw new Error("빈 범위식");
  for (const { pat, why } of REJECT) {
    if (raw.includes(pat)) throw new Error(`미지원 범위 문법 "${pat}": ${why} (11.3)`);
  }
  const tokens = raw.split(/\s+/);
  const comparators: Comparator[] = [];
  for (const tok of tokens) {
    const m = COMPARATOR.exec(tok);
    if (!m) {
      // x-range(1.2.x / 1.2.* / 1) 및 기타 미지원 형태를 여기서 거부.
      throw new Error(`미지원/유효하지 않은 비교자 "${tok}" — x-range·부분버전 거부 (11.3)`);
    }
    comparators.push({ op: (m[1] ?? "=") as Op, version: parseVersion(m[2]!) });
  }
  return comparators;
}

function satisfiesComparator(v: SemVer, c: Comparator): boolean {
  const cmp = compare(v, c.version);
  switch (c.op) {
    case ">=": return cmp >= 0;
    case "<=": return cmp <= 0;
    case ">": return cmp > 0;
    case "<": return cmp < 0;
    case "=": return cmp === 0;
  }
}

/**
 * 버전이 범위(AND 결합)를 만족하는가.
 * 프리릴리스 앱 버전은 기본 미매칭. matchPrerelease=true면 평가하되,
 * node-semver 관례처럼 '같은 (major,minor,patch) 튜플에 프리릴리스를 명시한 비교자가
 * 하나라도 있을 때만' 프리릴리스가 범위에 들어올 수 있게 한다.
 */
export function satisfies(
  version: SemVer,
  comparators: readonly Comparator[],
  opts: { matchPrerelease?: boolean } = {},
): boolean {
  if (isPrerelease(version)) {
    if (!opts.matchPrerelease) return false;
    const tupleMatch = comparators.some(
      (c) =>
        isPrerelease(c.version) &&
        c.version.major === version.major &&
        c.version.minor === version.minor &&
        c.version.patch === version.patch,
    );
    if (!tupleMatch) return false;
  }
  return comparators.every((c) => satisfiesComparator(version, c));
}

/** 편의: 문자열 버전 + 문자열 범위. */
export function versionInRange(
  version: string,
  range: string,
  opts?: { matchPrerelease?: boolean },
): boolean {
  return satisfies(parseVersion(version), parseRange(range), opts);
}

File: sdks/ios/Sources/RynL10n/SemVer.swift
import Foundation

/// node-semver 부분집합 — 기획서 11.3.
/// 지원: 비교자 `>= > <= < =` + 공백 AND. 거부: `||`·`^`·`~`·x-range·hyphen-range.
public struct SemVer: Equatable, Sendable {
    public let major: Int
    public let minor: Int
    public let patch: Int
    public let prerelease: [PreId]

    public enum PreId: Equatable, Sendable {
        case num(Int)
        case text(String)
    }

    public var isPrerelease: Bool { !prerelease.isEmpty }
}

public enum SemVerError: Error, Equatable {
    case invalidVersion(String)
    case unsupportedRange(String)
}

public enum SemVerParser {
    public static func parseVersion(_ input: String) throws -> SemVer {
        let trimmed = input.trimmingCharacters(in: .whitespaces)
        // major.minor.patch(-prerelease)?(+build)?
        let pattern = #"^(\d+)\.(\d+)\.(\d+)(?:-([0-9A-Za-z.-]+))?(?:\+[0-9A-Za-z.-]+)?$"#
        guard let m = firstMatch(trimmed, pattern) else { throw SemVerError.invalidVersion(input) }
        let major = Int(m[1]!)!, minor = Int(m[2]!)!, patch = Int(m[3]!)!
        var pre: [SemVer.PreId] = []
        if let p = m[4], !p.isEmpty {
            pre = p.split(separator: ".").map { id in
                if let n = Int(id), String(n) == id { return .num(n) }
                return .text(String(id))
            }
        }
        return SemVer(major: major, minor: minor, patch: patch, prerelease: pre)
    }

    public static func compare(_ a: SemVer, _ b: SemVer) -> Int {
        if a.major != b.major { return a.major < b.major ? -1 : 1 }
        if a.minor != b.minor { return a.minor < b.minor ? -1 : 1 }
        if a.patch != b.patch { return a.patch < b.patch ? -1 : 1 }
        let ap = a.prerelease, bp = b.prerelease
        if ap.isEmpty && bp.isEmpty { return 0 }
        if ap.isEmpty { return 1 }
        if bp.isEmpty { return -1 }
        for i in 0..<min(ap.count, bp.count) {
            switch (ap[i], bp[i]) {
            case let (.num(x), .num(y)): if x != y { return x < y ? -1 : 1 }
            case (.num, .text): return -1
            case (.text, .num): return 1
            case let (.text(x), .text(y)): if x != y { return x < y ? -1 : 1 }
            }
        }
        if ap.count != bp.count { return ap.count < bp.count ? -1 : 1 }
        return 0
    }

    public enum Op: String { case gte = ">=", lte = "<=", gt = ">", lt = "<", eq = "=" }
    public struct Comparator: Sendable { public let op: String; public let version: SemVer }

    public static func parseRange(_ input: String) throws -> [Comparator] {
        let raw = input.trimmingCharacters(in: .whitespaces)
        if raw.isEmpty { throw SemVerError.unsupportedRange("빈 범위식") }
        for pat in ["||", "^", "~", " - "] where raw.contains(pat) {
            throw SemVerError.unsupportedRange("미지원 범위 문법 \"\(pat)\"")
        }
        let tokens = raw.split(whereSeparator: { $0 == " " }).map(String.init)
        var comparators: [Comparator] = []
        let compPattern = #"^(>=|<=|>|<|=)?(\d+\.\d+\.\d+(?:-[0-9A-Za-z.-]+)?)$"#
        for tok in tokens {
            guard let m = firstMatch(tok, compPattern) else {
                throw SemVerError.unsupportedRange("미지원/유효하지 않은 비교자 \"\(tok)\"")
            }
            let op = m[1] ?? "="
            comparators.append(Comparator(op: op, version: try parseVersion(m[2]!)))
        }
        return comparators
    }

    public static func satisfies(_ version: SemVer, _ comparators: [Comparator], matchPrerelease: Bool = false) -> Bool {
        if version.isPrerelease {
            if !matchPrerelease { return false }
            let tupleMatch = comparators.contains { c in
                c.version.isPrerelease && c.version.major == version.major
                    && c.version.minor == version.minor && c.version.patch == version.patch
            }
            if !tupleMatch { return false }
        }
        return comparators.allSatisfy { c in
            let cmp = compare(version, c.version)
            switch c.op {
            case ">=": return cmp >= 0
            case "<=": return cmp <= 0
            case ">": return cmp > 0
            case "<": return cmp < 0
            default: return cmp == 0
            }
        }
    }

    public static func versionInRange(_ version: String, _ range: String, matchPrerelease: Bool = false) throws -> Bool {
        satisfies(try parseVersion(version), try parseRange(range), matchPrerelease: matchPrerelease)
    }
}

/// 캡처 그룹을 옵셔널 문자열 배열로 반환하는 정규식 헬퍼(0=전체, 1..=그룹).
func firstMatch(_ s: String, _ pattern: String) -> [String?]? {
    guard let re = try? NSRegularExpression(pattern: pattern) else { return nil }
    let range = NSRange(s.startIndex..<s.endIndex, in: s)
    guard let m = re.firstMatch(in: s, range: range) else { return nil }
    var groups: [String?] = []
    for i in 0..<m.numberOfRanges {
        let r = m.range(at: i)
        if r.location == NSNotFound { groups.append(nil) }
        else if let rr = Range(r, in: s) { groups.append(String(s[rr])) }
        else { groups.append(nil) }
    }
    return groups
}
